dll/win32/shlwapi/autocomp.cpp

```diff
@@ -0,0 +1,399 @@
+        }
+    }
+
+    RegCloseKey(hKey);
+}
+
+static BOOL
+AutoComplete_AdaptFlags(HWND hwndEdit, LPDWORD pdwACO, LPDWORD pdwSHACF)
+{
+    static const LPCWSTR s_pszAutoComplete =
+        L"Software\\Microsoft\\Internet Explorer\\AutoComplete";
+    WCHAR szValue[8];
+
+    DWORD dwSHACF = *pdwSHACF, dwACO = 0;
+    if (dwSHACF == SHACF_DEFAULT)
+        dwSHACF = SHACF_FILESYSTEM | SHACF_URLALL;
+
+    if (dwSHACF & SHACF_AUTOAPPEND_FORCE_OFF)
+    {
+        // do nothing
+    }
+    else if (dwSHACF & SHACF_AUTOAPPEND_FORCE_ON)
+    {
+        dwACO |= ACO_AUTOAPPEND;
+    }
+    else
+    {
+        DWORD dwType, cbValue = sizeof(szValue);
+        if (SHGetValueW(HKEY_CURRENT_USER, s_pszAutoComplete, L"Append Completion",
+                        &dwType, szValue, &cbValue) != ERROR_SUCCESS ||
+            dwType != REG_SZ || _wcsicmp(szValue, L"no") != 0)
+        {
+            dwACO |= ACO_AUTOSUGGEST;
+        }
+    }
+
+    if (dwSHACF & SHACF_AUTOSUGGEST_FORCE_OFF)
+    {
+        // do nothing
+    }
+    else if (dwSHACF & SHACF_AUTOSUGGEST_FORCE_ON)
+    {
+        dwACO |= ACO_AUTOSUGGEST;
+    }
+    else
+    {
+        DWORD dwType, cbValue = sizeof(szValue);
+        if (SHGetValueW(HKEY_CURRENT_USER, s_pszAutoComplete, L"AutoSuggest",
+                        &dwType, szValue, &cbValue) != ERROR_SUCCESS ||
+            dwType != REG_SZ || _wcsicmp(szValue, L"no") != 0)
+        {
+            dwACO |= ACO_AUTOSUGGEST;
+        }
+    }
+
+    if (dwSHACF & SHACF_USETAB)
+        dwACO |= ACO_USETAB;
+
+    if (GetWindowLongPtr(hwndEdit, GWL_EXSTYLE) & WS_EX_RTLREADING)
+        dwACO |= ACO_RTLREADING;
+
+    if (!(dwACO & (ACO_AUTOSUGGEST | ACO_AUTOAPPEND)))
+    {
+        ERR("dwACO: 0x%lX\n", dwACO);
+        return FALSE;
+    }
+
+    *pdwACO = dwACO;
+    *pdwSHACF = dwSHACF;
+    return TRUE;
+}
+
+/*************************************************************************
+ *      SHAutoComplete  	[SHLWAPI.@]
+ *
+ * Enable auto-completion for an edit control.
+ *
+ * PARAMS
+ *  hwndEdit [I] Handle of control to enable auto-completion for
+ *  dwFlags  [I] SHACF_ flags from "shlwapi.h"
+ *
+ * RETURNS
+ *  Success: S_OK. Auto-completion is enabled for the control.
+ *  Failure: An HRESULT error code indicating the error.
+ */
+HRESULT WINAPI SHAutoComplete(HWND hwndEdit, DWORD dwFlags)
+{
+    TRACE("SHAutoComplete(%p, 0x%lX)\n", hwndEdit, dwFlags);
+
+    DWORD dwACO;
+    if (!AutoComplete_AdaptFlags(hwndEdit, &dwACO, &dwFlags))
+        return S_OK;
+
+    CComPtr<IAutoComplete2> pAC2;
+    const DWORD dwClsCtx = CLSCTX_INPROC_HANDLER | CLSCTX_INPROC_SERVER | CLSCTX_LOCAL_SERVER;
+    HRESULT hr = CoCreateInstance(CLSID_AutoComplete, NULL, dwClsCtx,
+                                  IID_IAutoComplete, (LPVOID *)&pAC2);
+    if (FAILED(hr))
+    {
+        ERR("CoCreateInstance(CLSID_AutoComplete) failed: 0x%lX\n", hr);
+        return hr;
+    }
+
+    CComPtr<IEnumString> pES;
```

Comment:
`CComObject<IEnumString> ES`
or `CComObject<IEnumString>* pES`

That implements the QueryInterface and family

Reply:
OK.